src/node_messaging.cc

```diff
@@ -537,36 +544,87 @@ std::unique_ptr<MessagePortData> MessagePort::Detach() {
 }
 
 
-void MessagePort::Send(Message&& message) {
-  Mutex::ScopedLock lock(*data_->sibling_mutex_);
-  if (data_->sibling_ == nullptr)
-    return;
-  data_->sibling_->AddToIncomingQueue(std::move(message));
-}
+Maybe<bool> MessagePort::PostMessage(Environment* env,
+                                     Local<Value> message_v,
+                                     Local<Value> transfer_v) {
+  auto isolate = env->isolate();
+  auto obj = object(isolate);
+  auto context = obj->CreationContext();
+
+  std::vector<MessagePort*> transferred_ports;
+
+  // Check if the MessagePort itself is being transferred.
+  if (transfer_v->IsArray()) {
+    Local<Array> transfer_list = transfer_v.As<Array>();
+    uint32_t length = transfer_list->Length();
+    for (uint32_t i = 0; i < length; ++i) {
+      Local<Value> entry;
+      if (!transfer_list->Get(context, i).ToLocal(&entry))
+        return Nothing<bool>();
+
+      if (env->message_port_constructor_template()->HasInstance(entry)) {
+        if (entry == obj) {
+          Local<Value> argv[] = {
+            FIXED_ONE_BYTE_STRING(isolate,
+                                  "Transfer list contains source port"),
+            FIXED_ONE_BYTE_STRING(isolate, "DataCloneError")
+          };
+          env->isolate()->ThrowException(
+              env->domexception_function()
+                  ->NewInstance(context, arraysize(argv), argv)
+                  .ToLocalChecked());
+          return Nothing<bool>();
+        }
+        MessagePort* port = Unwrap<MessagePort>(entry.As<Object>());
+        CHECK_NE(port, nullptr);
+        transferred_ports.push_back(port);
+      }
+    }
+  }
 
-void MessagePort::Send(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Local<Context> context = object(env->isolate())->CreationContext();
   Message msg;
-  if (msg.Serialize(env, context, args[0], args[1])
-          .IsNothing()) {
-    return;
+
+  // Per spec, we need to serialize the input message even if the MessagePort
```

Comment:
Can we avoid `Per spec` here sine we are not making a commitment to follow any spec yet with regards to worker. I'd prefer "to align with web worker behavior, we serialize ..." - is that OK?

Reply:
The `MessagePort`/`MessageChannel` API is pretty close to the [Messaging part of the HTML spec](https://html.spec.whatwg.org/multipage/web-messaging.html), and not related to Workers per se.

At least I wrote it with that spec in mind and using it as a reference document.

Reply:
While workers are a different matter, the channel messaging API is in my opinion simple enough so that almost* full compatibility with HTML is a worthwhile goal.

\* `EventEmitter` notwithstanding.